// console-test-helpers@0.3.3/lib/console-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const CONSOLE_WRITE_METHODS = [
    'log',
    'debug',
    'info',
    'warn',
    'error',
    'dir',
    'trace',
    'group',
    'groupCollapsed',
    'groupEnd',
    'table',
];
class ConsoleState {
    constructor() {
        this.state = new Map();
    }
    record(method, message) {
        var _a;
        this.verifyMethod(method);
        if (!this.state.has(method)) {
            this.state.set(method, []);
        }
        let methodState = this.state.get(method);
        (_a = methodState) === null || _a === void 0 ? void 0 : _a.push(message);
    }
    getState(method) {
        var _a;
        this.verifyMethod(method);
        return ((_a = this.state.get(method)) === null || _a === void 0 ? void 0 : _a.join('\n')) || '';
    }
    log() {
        return this.getState('log');
    }
    debug() {
        return this.getState('debug');
    }
    info() {
        return this.getState('info');
    }
    warn() {
        return this.getState('warn');
    }
    error() {
        return this.getState('error');
    }
    dir() {
        return this.getState('dir');
    }
    group() {
        return this.getState('group');
    }
    groupCollapsed() {
        return this.getState('groupCollapsed');
    }
    groupEnd() {
        return this.getState('groupEnd');
    }
    table() {
        return this.getState('table');
    }
    verifyMethod(method) {
        if (!CONSOLE_WRITE_METHODS.includes(method)) {
            throw new Error(`The supplied method '${method}' is not supported`);
        }
    }
}
exports.default = ConsoleState;
